In [1]:
import pandas as pd
import matplotlib.pyplot as plt
from ase.io import read
from ase import Atoms
from pathlib import Path

O presente Notebook tem o intuito de analisar os dados de cálculos DFT para a determinação semi-empírica dos parâmetros de Hubbard, utilizando pseudopotenciais da biblioteca pseudo-dojo de norma conservada, na aproximação GGA-PBE.

In [2]:
# Reading data with Atomic Simulation Environment and Pathlib.
fpaths = "/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/vcrelax-files"
relax_paths: list[Path] = [f for f in Path(fpaths).iterdir() if f.suffix == ".out"]
relax_outputs: list[Atoms] = [read(fpath) for fpath in relax_paths]

In [3]:
df = pd.DataFrame(
    {
        "fpath": [f for f in relax_paths],
        "fname": [f.name for f in relax_paths],
        "a": [atom.cell[0][0] for atom in relax_outputs],
        "c": [atom.cell[2][2] for atom in relax_outputs],
    }
)
print("DataFrama com dados obtidos do Atomic Simulation Environment")
display(df)
print("DataFrama com dados obtidos através de arquivo processado com bash\n[extract_bandgaps.sh => bandgaps.dat]")

df_gap = pd.read_csv("/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/bandgaps.dat", sep=" ").reset_index(names="fname")
display(df_gap)

# Merge nos DataFrames pela coluna em comum "fname".
print("Merge nos DataFrames pela coluna em comum 'fname'.")
df = df.merge(right=df_gap, how="left").sort_values(by="fname").reset_index(drop=True)
display(df)


DataFrama com dados obtidos do Atomic Simulation Environment


,fpath,fname,a,c
0,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,ZnO-PBEU-9.50_7.00.out,3.246229,5.193781
1,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,ZnO-PBEU-5.00_5.00.out,3.253137,5.211166
2,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,ZnO-PBEU-6.50_8.00.out,3.238779,5.182282
3,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,ZnO-PBEU-8.00_5.00.out,3.255841,5.213483
4,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,ZnO-PBEU-4.00_8.50.out,3.234023,5.175905
...,...,...,...,...
284,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,ZnO-PBEU-7.50_7.50.out,3.242095,5.187599
285,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,ZnO-PBEU-11.50_8.00.out,3.242758,5.185719
286,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,ZnO-PBEU-7.00_8.00.out,3.239221,5.182652
287,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,ZnO-PBEU-8.50_4.00.out,3.262159,5.225436


DataFrama com dados obtidos através de arquivo processado com bash
[extract_bandgaps.sh => bandgaps.dat]


,fname,highest_occupied,lowest_unoccupied,Egap
0,ZnO-PBEU-10.00_10.00.out,7.0464,11.1901,4.1437
1,ZnO-PBEU-10.00_10.50.out,6.9622,11.2426,4.2804
2,ZnO-PBEU-10.00_11.00.out,6.8783,11.2941,4.4158
3,ZnO-PBEU-10.00_11.50.out,6.7947,11.3444,4.5497
4,ZnO-PBEU-10.00_12.00.out,6.7116,11.3938,4.6822
...,...,...,...,...
284,ZnO-PBEU-9.50_7.50.out,7.4991,10.9135,3.4144
285,ZnO-PBEU-9.50_8.00.out,7.4170,10.9724,3.5554
286,ZnO-PBEU-9.50_8.50.out,7.3345,11.0300,3.6955
287,ZnO-PBEU-9.50_9.00.out,7.2520,11.0863,3.8343


Merge nos DataFrames pela coluna em comum 'fname'.


,fpath,fname,a,c,highest_occupied,lowest_unoccupied,Egap
0,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,ZnO-PBEU-10.00_10.00.out,3.232534,5.167752,7.0464,11.1901,4.1437
1,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,ZnO-PBEU-10.00_10.50.out,3.230404,5.163852,6.9622,11.2426,4.2804
2,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,ZnO-PBEU-10.00_11.00.out,3.228332,5.160034,6.8783,11.2941,4.4158
3,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,ZnO-PBEU-10.00_11.50.out,3.226313,5.156344,6.7947,11.3444,4.5497
4,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,ZnO-PBEU-10.00_12.00.out,3.224341,5.152755,6.7116,11.3938,4.6822
...,...,...,...,...,...,...,...
284,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,ZnO-PBEU-9.50_7.50.out,3.243721,5.189010,7.4991,10.9135,3.4144
285,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,ZnO-PBEU-9.50_8.00.out,3.241276,5.184405,7.4170,10.9724,3.5554
286,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,ZnO-PBEU-9.50_8.50.out,3.238901,5.179962,7.3345,11.0300,3.6955
287,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,ZnO-PBEU-9.50_9.00.out,3.236589,5.175649,7.2520,11.0863,3.8343


In [4]:
# Filtrando pelo Band Gap Experimental e criando coluna da razão c/a para comparação experimental.
df_sorted_gaps = df.query("3.3 <= Egap <= 3.44").sort_values(by="Egap").reset_index(drop=True)
df_sorted_gaps["c/a"] = df_sorted_gaps["c"] / df_sorted_gaps["a"]

display(df_sorted_gaps)

,fpath,fname,a,c,highest_occupied,lowest_unoccupied,Egap,c/a
0,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,ZnO-PBEU-6.00_8.00.out,3.238325,5.181908,7.7075,11.0114,3.3039,1.600182
1,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,ZnO-PBEU-10.50_7.00.out,3.246985,5.194461,7.5209,10.8431,3.3222,1.599780
2,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,ZnO-PBEU-8.00_7.50.out,3.242516,5.187960,7.6058,10.9294,3.3236,1.599979
3,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,ZnO-PBEU-5.00_8.50.out,3.235012,5.176716,7.7483,11.0812,3.3329,1.600215
4,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,ZnO-PBEU-4.00_9.00.out,3.231715,5.171612,7.8127,11.1506,3.3379,1.600269
5,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,ZnO-PBEU-11.00_7.00.out,3.247350,5.194790,7.4929,10.8382,3.3453,1.599701
6,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,ZnO-PBEU-6.50_8.00.out,3.238779,5.182282,7.6592,11.0054,3.3462,1.600073
7,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,ZnO-PBEU-8.50_7.50.out,3.242928,5.188315,7.5685,10.9240,3.3555,1.599886
8,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,ZnO-PBEU-11.50_7.00.out,3.247709,5.195112,7.4662,10.8334,3.3672,1.599624
9,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,ZnO-PBEU-5.50_8.50.out,3.235485,5.177083,7.6921,11.0751,3.3830,1.600095


Verificação de convergência do BFGS com __Bash__:


```bash
for f in *.out; do
    grep -q "bfgs converged" "$f" \
        && echo "$f : Sim" \
        || echo "$f : Não"
done
```


In [5]:
def convergiu(fpath: Path | str) -> str:
    """Verificar a convergência do BFGS"""
    try:
        text = Path(fpath).read_text(errors="ignore")
        status = "Sim" if "bfgs converged" in text.lower() else "Não"
        return status
    except FileNotFoundError:
        "FN"

In [6]:
df_sorted_gaps = df_sorted_gaps.reindex(columns=["fname","a","c","c/a", "lowest_unoccupied", "highest_occupied", "Egap","fpath"])
df_sorted_gaps

,fname,a,c,c/a,lowest_unoccupied,highest_occupied,Egap,fpath
0,ZnO-PBEU-6.00_8.00.out,3.238325,5.181908,1.600182,11.0114,7.7075,3.3039,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...
1,ZnO-PBEU-10.50_7.00.out,3.246985,5.194461,1.599780,10.8431,7.5209,3.3222,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...
2,ZnO-PBEU-8.00_7.50.out,3.242516,5.187960,1.599979,10.9294,7.6058,3.3236,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...
3,ZnO-PBEU-5.00_8.50.out,3.235012,5.176716,1.600215,11.0812,7.7483,3.3329,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...
4,ZnO-PBEU-4.00_9.00.out,3.231715,5.171612,1.600269,11.1506,7.8127,3.3379,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...
5,ZnO-PBEU-11.00_7.00.out,3.247350,5.194790,1.599701,10.8382,7.4929,3.3453,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...
6,ZnO-PBEU-6.50_8.00.out,3.238779,5.182282,1.600073,11.0054,7.6592,3.3462,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...
7,ZnO-PBEU-8.50_7.50.out,3.242928,5.188315,1.599886,10.9240,7.5685,3.3555,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...
8,ZnO-PBEU-11.50_7.00.out,3.247709,5.195112,1.599624,10.8334,7.4662,3.3672,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...
9,ZnO-PBEU-5.50_8.50.out,3.235485,5.177083,1.600095,11.0751,7.6921,3.3830,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...


In [7]:
df_sorted_gaps["Convergiu"] = df_sorted_gaps["fpath"].apply(convergiu)
display(df_sorted_gaps)

,fname,a,c,c/a,lowest_unoccupied,highest_occupied,Egap,fpath,Convergiu
0,ZnO-PBEU-6.00_8.00.out,3.238325,5.181908,1.600182,11.0114,7.7075,3.3039,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,Sim
1,ZnO-PBEU-10.50_7.00.out,3.246985,5.194461,1.599780,10.8431,7.5209,3.3222,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,Sim
2,ZnO-PBEU-8.00_7.50.out,3.242516,5.187960,1.599979,10.9294,7.6058,3.3236,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,Sim
3,ZnO-PBEU-5.00_8.50.out,3.235012,5.176716,1.600215,11.0812,7.7483,3.3329,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,Sim
4,ZnO-PBEU-4.00_9.00.out,3.231715,5.171612,1.600269,11.1506,7.8127,3.3379,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,Sim
5,ZnO-PBEU-11.00_7.00.out,3.247350,5.194790,1.599701,10.8382,7.4929,3.3453,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,Sim
6,ZnO-PBEU-6.50_8.00.out,3.238779,5.182282,1.600073,11.0054,7.6592,3.3462,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,Sim
7,ZnO-PBEU-8.50_7.50.out,3.242928,5.188315,1.599886,10.9240,7.5685,3.3555,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,Sim
8,ZnO-PBEU-11.50_7.00.out,3.247709,5.195112,1.599624,10.8334,7.4662,3.3672,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,Sim
9,ZnO-PBEU-5.50_8.50.out,3.235485,5.177083,1.600095,11.0751,7.6921,3.3830,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,Sim


<font color="lightblue" size=5> Constantes de Rede</font>

In [8]:
mean_a = df_sorted_gaps['a'].mean()
mean_c = df_sorted_gaps['c'].mean()

print(f"Valor médio de a: {mean_a:.4f}")
print(f"Valor médio de c: {mean_c:.4f}")

Valor médio de a: 3.2405
Valor médio de c: 5.1846


In [9]:
df_sorted_gaps.sort_values(by="a", ascending=False).head(5)

,fname,a,c,c/a,lowest_unoccupied,highest_occupied,Egap,fpath,Convergiu
12,ZnO-PBEU-12.00_7.00.out,3.248059,5.195433,1.599550,10.8287,7.4407,3.3880,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,Sim
8,ZnO-PBEU-11.50_7.00.out,3.247709,5.195112,1.599624,10.8334,7.4662,3.3672,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,Sim
5,ZnO-PBEU-11.00_7.00.out,3.247350,5.194790,1.599701,10.8382,7.4929,3.3453,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,Sim
1,ZnO-PBEU-10.50_7.00.out,3.246985,5.194461,1.599780,10.8431,7.5209,3.3222,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,Sim
14,ZnO-PBEU-9.50_7.50.out,3.243721,5.189010,1.599709,10.9135,7.4991,3.4144,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,Sim


In [10]:
df_sorted_gaps.sort_values(by="c", ascending=False).head(5)

,fname,a,c,c/a,lowest_unoccupied,highest_occupied,Egap,fpath,Convergiu
12,ZnO-PBEU-12.00_7.00.out,3.248059,5.195433,1.599550,10.8287,7.4407,3.3880,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,Sim
8,ZnO-PBEU-11.50_7.00.out,3.247709,5.195112,1.599624,10.8334,7.4662,3.3672,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,Sim
5,ZnO-PBEU-11.00_7.00.out,3.247350,5.194790,1.599701,10.8382,7.4929,3.3453,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,Sim
1,ZnO-PBEU-10.50_7.00.out,3.246985,5.194461,1.599780,10.8431,7.5209,3.3222,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,Sim
14,ZnO-PBEU-9.50_7.50.out,3.243721,5.189010,1.599709,10.9135,7.4991,3.4144,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,Sim


In [11]:
df_sorted_gaps.sort_values(by="c/a", ascending=True).head(5)


,fname,a,c,c/a,lowest_unoccupied,highest_occupied,Egap,fpath,Convergiu
12,ZnO-PBEU-12.00_7.00.out,3.248059,5.195433,1.599550,10.8287,7.4407,3.3880,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,Sim
8,ZnO-PBEU-11.50_7.00.out,3.247709,5.195112,1.599624,10.8334,7.4662,3.3672,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,Sim
5,ZnO-PBEU-11.00_7.00.out,3.247350,5.194790,1.599701,10.8382,7.4929,3.3453,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,Sim
14,ZnO-PBEU-9.50_7.50.out,3.243721,5.189010,1.599709,10.9135,7.4991,3.4144,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,Sim
1,ZnO-PBEU-10.50_7.00.out,3.246985,5.194461,1.599780,10.8431,7.5209,3.3222,/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/v...,Sim


Ao que tudo indica eu PODERIA, escolher o mesmo valor de $U_d$ e $U_p$ que eu escolhi para o LDA+U $(12.0, 7.0)$
<br>
<br>
<br>

<div align="center">

| Parâmetro de Rede | LDA | LDA+U | GGA | GGA+U |
|-------------------|-----|-------|-----|-------|
| a | 3.168509 | 3.168506 | 3.281166 | 3.248059 |
| c | 5.065865 | 5.065878 |5.274064 | 5.195432 |
| c/a | 1.598817 | 1.598822 | 1.607375 | 1.599550 | 
</div>



In [12]:
from ase.io import read


In [13]:
atoms = read("/home/jvc/ZnO_database/DFT_PBEU/01_RelaxPBEU/vcrelax-files/ZnO-PBEU-12.00_7.00.out")
unit_cell = atoms.cell

print(unit_cell)

A = atoms.cell[0][0]
C = atoms.cell[2][2]

print(f"Razão c/a = {C/A:.6f}")

Cell([[3.2480592192567976, 0.0, 0.0], [-1.6240296096283988, 2.8129017957827727, 0.0], [0.0, 0.0, 5.195432544778627]])
Razão c/a = 1.599550
